### PR TITLE
Hakija readonly question group

### PR DESCRIPTION
**Pull request by heeboA**

### CODE REVIEW

File: src/cljs/ataru/application_common/application_field_common.cljs [hunk 1]
```diff
@@ -77,6 +80,11 @@
 
       :else value-or-koodi-uri)))
 
+(defn group-spacer
+  [index]
+  [^{:key (str "spacer-" index)}
```
Comment (elahti) [inline src/cljs/ataru/application_common/application_field_common.cljs]: Isn't the outer vector in the return value here unnecessary? Just returning

```clojure
^{:key (str "spacer-" index)}
[:div.application__question-group-spacer]
```

should be enough?

File: src/cljs/ataru/hakija/application_form_components.cljs [hunk 2]
```diff
@@ -288,11 +289,9 @@
                              [render-field child :idx idx])
                            children)
                         (when (< idx (dec row-count))
-                          [^{:key (str "spacer-" row-count)}
-                            [:div.application__question-group-spacer]])))
+                          (group-spacer idx))))
```
Comment (elahti) [inline src/cljs/ataru/hakija/application_form_components.cljs]: Should a new component just be created here instead:  `[group-spacer idx]`.

File: src/cljs/ataru/hakija/hakija_readonly.cljs [hunk 3]
```diff
@@ -18,40 +18,60 @@
...
-        (cond (and (vector? (:values answer))
-                   (every? vector? (:values answer)))
-              (into [:ul.application__form-field-list] (map multi-values->:li (:values answer)))
-
-              (and (vector? (:values answer))
-                   (some (comp not vector?) (:values answer)))
-              (into [:ul.application__form-field-list] (multi-values->:li (:values answer)))
+       [:div.application__readonly-text
+        (cond (and (vector? values)
+                   (every? vector? values)
```
Comment (elahti) [inline src/cljs/ataru/hakija/hakija_readonly.cljs]: Should a `(not (empty? values))` check be in the `and` clause here too? Invoking

```clojure
(def values [])
(and (vector? values)
     (every? vector? values))
```

will return `true`. Should also note that there are a lot of other places to be fixed regarding this.

Comment (elahti) [reply src/cljs/ataru/hakija/hakija_readonly.cljs]: Maybe we should have an util function to actually check this and then use that function everywhere?

File: src/cljs/ataru/hakija/hakija_readonly.cljs [hunk 4]
```diff
@@ -18,40 +18,60 @@
...
+                   (not= 1 (count values)))
+              (into [:ul.application__form-field-list]
+                    (map #(multi-values->:li field-descriptor lang %) values))
+
+              (and (vector? values)
+                   (some (comp not vector?) values)
```
Comment (elahti) [inline src/cljs/ataru/hakija/hakija_readonly.cljs]: I'm thinking there might be a possibility this does return false positives on some cases. Please see my previous comment above.

File: src/cljs/ataru/hakija/hakija_readonly.cljs [hunk 5]
```diff
@@ -76,12 +96,36 @@
...
+               (map
+                 (fn [group-index]
+                   (concat
+                     (map-indexed
+                       (fn [i [id [field-descriptor]]]
+                         ^{:key (str id "-" (+ group-index i))}
+                         [field field-descriptor application lang group-index])
+                       fields)
+                     (when (< group-index (dec groups-amount))
+                       (group-spacer group-index))))
```
Comment (elahti) [inline src/cljs/ataru/hakija/hakija_readonly.cljs]: Should a new component just be created here instead: `[group-spacer idx]`.

File: resources/less/hakija.less [hunk 6]
```diff
@@ -314,7 +322,7 @@
 }
 
 .application__question-group-spacer + .application__form-field,
-.application__question-group-wrapper-contents .application__form-field:first-child {
+.application__question-group-wrapper-contents div.application__form-field:nth-of-type(1) {
```
[the remark this answers is not in the document]
Comment (heeboA) [reply resources/less/hakija.less]: There is a <p.application__form-field> as first child in readonly view, so yeah either only div or with class, maybe safer with class if someone adds something there.

File: resources/less/hakija.less [hunk 7]
```diff
@@ -733,17 +741,26 @@ textarea.application__form-text-area__size-large {
...
 }
 
 th.application__readonly-adjacent--header {
   font-weight: 600;
   min-width: 33%;
 }
 
+.application__readonly-text {
+  -ms-word-wrap: break-word;
+  word-wrap: break-word;
```
[the remark this answers is not in the document]
Comment (heeboA) [reply resources/less/hakija.less]: Removed alltogether, i had a long string without whitespace as test text, which seems unlikely.

File: src/cljs/ataru/application_common/application_field_common.cljs [hunk 8]
```diff
@@ -50,13 +50,16 @@
...
   (let [key                (answer-key field-descriptor)
-        value-or-koodi-uri (:value (get (:answers application) key))
+        value-or-koodi-uri (if question-group-index
+                             (-> application :answers key :value (nth question-group-index))
+                             (-> application :answers key :value))
         is-koodisto?       (contains? field-descriptor :koodisto-source)
         split-values       (cond-> value-or-koodi-uri
                                    (and is-koodisto? (string? value-or-koodi-uri))
                                    (clojure.string/split #"\s*,\s*"))]
+    ;(println value-or-koodi-uri)
```
[the remark this answers is not in the document]
Comment (heeboA) [reply src/cljs/ataru/application_common/application_field_common.cljs]: yes

File: src/cljs/ataru/application_common/application_field_common.cljs (hunk 1 above)
Comment (heeboA) [reply src/cljs/ataru/application_common/application_field_common.cljs]: Yeah, fixed by the next comment!

File: src/cljs/ataru/hakija/application_form_components.cljs (hunk 2 above)
Comment (heeboA) [reply src/cljs/ataru/hakija/application_form_components.cljs]: Indeed, still gotta figure out re-frame :)

File: src/cljs/ataru/application_common/application_field_common.cljs (hunk 1 above)
Comment (heeboA) [reply src/cljs/ataru/application_common/application_field_common.cljs]: Yeah!

File: src/cljs/ataru/hakija/application_form_components.cljs (hunk 2 above)
Comment (heeboA) [reply src/cljs/ataru/hakija/application_form_components.cljs]: Yeah works:+1:

File: src/cljs/ataru/hakija/hakija_readonly.cljs (hunk 3 above)
Comment (heeboA) [reply src/cljs/ataru/hakija/hakija_readonly.cljs]: In that case will (map) over empty so no problem.

**Pull request merged**